sup/gpu.cpp

```diff
@@ -0,0 +1,54 @@
+#include <vector>
+
+#include <arbor/version.hpp>
+
+#ifdef ARB_GPU_ENABLED
+#include <cuda_runtime.h>
+#endif
+
+#ifdef ARB_MPI_ENABLED
+#include <mpi.h>
+#endif
+
```

Comment:
This code doesn't depend upon Arbor having been built with MPI or CUDA or not, but rather on the current configuration environment.

When we split `sup` up into a local-for-us library and an installed-for-general-consumption component, the latter will have to have similar CMake guff that the arbor library itself has to make this all hang together; in the meantime, we can hand-write the CMake stuff for `sup` that handles the `ARB_WITH_MPI` and `ARB_WITH_CUDA` stuff as required.

Reply:
I think I see what you are saying.

You are suggesting another set of `#defines`, e.g. `SUP_WITH_MPI`, `SUP_WITH_CUDA` from CMake?
We can't use the `ARB_WITH*` defines, because they shouldn't go past the public/private split.

We always want consistency between sup and arbor, i.e. if Arbor was compiled with MPI support, then sup must have matching MPI support.
The `libsup` installed with `libarbor` must be consistent, and the `libsup` and `libarbor` "used locally" for tests and the like must also always be consistent.

So `ARB_MPI_ENABLED` and `SUP_WITH_MPI` must always agree. Which makes me wonder if it isn't just simpler to use the one, even if it isn't a perfect split in theory.

To play the devil's advocate: two arguments for keeping it this way
* it is actually less confusing for users.
* one single variable, `ARB_XYZ_ENABLED`, has to be configured in CMake for every feature, instead of two (and the less CMake wrangling, the better).

Reply:
The `ARB_WITH_foo` defines are provided by the CMake infrastructure, and reflect the configuration environment. They are never exposed to user code, but there is no reason why they cannot also be used in the `sup` sources (obviously not the headers).

I can't see any need for the `sup` code to provide `arbor`-like feature macros.

The CMake guff will come from splitting  MPI and CUDA dependencies out from `arbor-private-deps` into an internal dependency target that will be a requirement for both `arbor` and `sup`.

Reply:
Addendum: the only reason why I've been advocating not using `ARB_WITH_foo` in our test sources is that it allows us to catch interface issues — it keeps us more honest when it come to corralling our configuration state.

Reply:
The point about minimising `ARB_WITH_foo` in the tests is a good one: we want to test the public interface whenever possible.

I am just trying to think about how to keep such `#ifdef MPI` out of the sup headers. When we have MPI support, there will be a header that exposes `find_private_gpu(MPI_Comm)`, which will obviously need to include `mpi.h`, like `sup/include/sup/with_mpi.hpp` currently does.

I would have to put the `find_private_gpu(MPI_Comm)` in a separate file to the `find_gpu()` (or whatever we call it, you get what I am saying).

Is this correct reasoning?

Reply:
The template option is a bit easier to implement, and is a pattern that we use elsewhere.
But compile errors are always preferable to linker errors.
I will think about it some more.